File: server/src/main/com/broll/gainea/server/core/actions/optional/AttackAction.java
package com.broll.gainea.server.core.actions.optional;

import com.broll.gainea.net.NT_Action_Attack;
import com.broll.gainea.net.NT_Reaction;
import com.broll.gainea.net.NT_Unit;
import com.broll.gainea.server.core.actions.AbstractActionHandler;
import com.broll.gainea.server.core.actions.ActionContext;
import com.broll.gainea.server.core.map.Location;
import com.broll.gainea.server.core.objects.BattleObject;
import com.broll.gainea.server.core.utils.PlayerUtils;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

public class AttackAction extends AbstractActionHandler<NT_Action_Attack, AttackAction.Context> {
    private final static Logger Log = LoggerFactory.getLogger(AttackAction.class);

    class Context extends ActionContext<NT_Action_Attack> {
        List<BattleObject> attackers;
        Location location;

        public Context(NT_Action_Attack action) {
            super(action);
        }
    }

    public Context attack(List<BattleObject> attackers, Location attackLocation) {
        NT_Action_Attack action = new NT_Action_Attack();
        action.units = attackers.stream().map(BattleObject::nt).toArray(NT_Unit[]::new);
        action.location = (short) attackLocation.getNumber();
        Context context = new Context(action);
        context.attackers = attackers;
        context.location = attackLocation;
        return context;
    }

    @Override
    public void handleReaction(Context context, NT_Action_Attack action, NT_Reaction reaction) {
        game.getProcessingCore().execute(() -> {
            Log.trace("Handle attack reaction");
            List<BattleObject> selectedAttackers = new ArrayList<>();
            Location from = null;
            for (int selection : reaction.options) {
                BattleObject attacker = context.attackers.get(selection);
                if (from == null) {
                    from = attacker.getLocation();
                } else if (attacker.getLocation() != from) {
                    continue;
                }
                selectedAttackers.add(attacker);
            }
            context.attackers.removeAll(selectedAttackers);
            if (!selectedAttackers.isEmpty()) {
                selectedAttackers.forEach(BattleObject::attacked);
                startFight(selectedAttackers, context.location);
            }
        });
    }

    private void startFight(List<BattleObject> attackers, Location attackLocation) {
        List<BattleObject> defenders = PlayerUtils.getHostileArmy(player, attackLocation);
        game.getBattleHandler().startBattle(attackers, defenders);
    }
}

File: server/src/main/com/broll/gainea/server/core/actions/optional/AttackAction.kt
package com.broll.gainea.server.core.actions.optional

import com.broll.gainea.net.NT_Action_Attack
import com.broll.gainea.net.NT_Reaction
import com.broll.gainea.server.core.actions.AbstractActionHandler
import com.broll.gainea.server.core.actions.ActionContext
import com.broll.gainea.server.core.map.Location
import com.broll.gainea.server.core.objects.Unit
import com.broll.gainea.server.core.utils.getHostileArmy
import org.slf4j.LoggerFactory

class AttackAction : AbstractActionHandler<NT_Action_Attack, AttackAction.Context>() {
    inner class Context(action: NT_Action_Attack, val attackers: MutableList<Unit>, val location: Location)
        : ActionContext<NT_Action_Attack>(action)

    fun attack(attackers: List<Unit>, attackLocation: Location): Context {
        val action = NT_Action_Attack()
        action.units = attackers.map { it.nt() }.toTypedArray()
        action.location = attackLocation.number.toShort()
        return Context(action, attackers.toMutableList(), attackLocation)
    }

    override fun handleReaction(context: Context, action: NT_Action_Attack, reaction: NT_Reaction) {
        game.processingCore.execute {
            Log.trace("Handle attack reaction")
            val selectedAttackers = mutableListOf<Unit>()
            var from: Location? = null
            reaction.options.map { selection -> context.attackers.find { it.id == selection } }.filterNotNull().forEach { unit ->
                if (from == null) {
                    from = unit.location
                }
                if (unit.location == from) {
                    selectedAttackers.add(unit)
                }
            }
            context.attackers.removeAll(selectedAttackers)
            if (selectedAttackers.isNotEmpty()) {
                selectedAttackers.forEach { it.attacked() }
                startFight(selectedAttackers, context.location)
            }
        }
    }

    private fun startFight(attackers: List<Unit>, attackLocation: Location) {
        val defenders = player.getHostileArmy(attackLocation)
        game.battleHandler.startBattle(attackers, defenders)
    }

    companion object {
        private val Log = LoggerFactory.getLogger(AttackAction::class.java)
    }
}
